cupyx/scipy/sparse/linalg/_solve.py

```diff
@@ -739,3 +740,238 @@ def _symOrtho(a, b):
+        maxiter = n * 5
+
+    istop = 0
+    itn = 0
+    Anorm = 0
+    Acond = 0
+    rnorm = 0
+    ynorm = 0
+
+    xtype = x.dtype
+
+    eps = cupy.finfo(xtype).eps
+
+    Ax = matvec(x)
+    r1 = b - Ax
+    y = psolve(r1)
+
+    beta1 = cupy.inner(r1, y)
+
+    if beta1 < 0:
+        raise ValueError('indefinite preconditioner')
+    elif beta1 == 0:
+        return x, 0
+
+    beta1 = cupy.sqrt(beta1)
+    beta1 = beta1.get().item()
+
+    if check:
+        # see if A is symmetric
+        if not _check_symmetric(A, Ax, x, eps):
+            raise ValueError('non-symmetric matrix')
+
+        # see if M is symmetric
+        if not _check_symmetric(M, y, r1, eps):
+            raise ValueError('non-symmetric preconditioner')
+
+    oldb = 0
+    beta = beta1
+    dbar = 0
+    epsln = 0
+    qrnorm = beta1
+    phibar = beta1
+    rhs1 = beta1
+    rhs2 = 0
+    tnorm2 = 0
+    gmax = 0
+    gmin = cupy.finfo(xtype).max
+    cs = -1
+    sn = 0
+    w = cupy.zeros(n, dtype=xtype)
+    w2 = cupy.zeros(n, dtype=xtype)
+    r2 = r1.copy()
+
+    while itn < maxiter:
+
+        itn += 1
+        s = 1.0 / beta
+        v = s * y
+
+        y = matvec(v)
+        y -= shift * v
+
+        if itn >= 2:
+            y -= (beta / oldb) * r1
+
+        alpha = cupy.inner(v, y)
+        y -= (alpha / beta) * r2
+        r1 = r2.copy()
+        r2 = y.copy()
+        y = psolve(r2)
+        oldb = beta
+        beta = cupy.inner(r2, y)
+        if beta < 0:
+            raise ValueError('non-symmetric matrix')
+        beta = cupy.sqrt(beta)
+        tnorm2 += alpha ** 2 + oldb ** 2 + beta ** 2
+
+        if itn == 1:
+            if beta / beta1 <= 10 * eps:
+                istop = -1
+
+        # Apply previous rotation Qk-1 to get
+        #   [deltak epslnk+1] = [cs  sn][dbark    0   ]
+        #   [gbar k dbar k+1]   [sn -cs][alfak betak+1].
+
+        oldeps = epsln
+        delta = cs * dbar + sn * alpha  # delta1 = 0         deltak
+        gbar = sn * dbar - cs * alpha  # gbar 1 = alfa1     gbar k
+        epsln = sn * beta  # epsln2 = 0         epslnk+1
+        dbar = - cs * beta  # dbar 2 = beta2     dbar k+1
+        root = numpy.linalg.norm([gbar, dbar])
+
+        # Compute the next plane rotation Qk
+
+        gamma = numpy.linalg.norm([gbar, beta])  # gammak
+        gamma = max(gamma, eps)
+        cs = gbar / gamma  # ck
+        sn = beta / gamma  # sk
+        phi = cs * phibar  # phik
+        phibar = sn * phibar  # phibark+1
+
+        # Update  x.
+
+        denom = 1.0 / gamma
+        w1 = w2.copy()
+        w2 = w.copy()
+        w = (v - oldeps * w1 - delta * w2) * denom
+        x += phi * w
+
+        # Go round again.
+
+        gmax = max(gmax, gamma)
+        gmin = min(gmin, gamma)
+        z = rhs1 / gamma
+        rhs1 = rhs2 - delta * z
+        rhs2 = - epsln * z
+
+        # Estimate various norms and test for convergence.
+
+        Anorm = cupy.sqrt(tnorm2)
+        ynorm = cupy.linalg.norm(x)
```

Comment:
Synchronizing `ynorm` to the host here would be better as it is used in scalar arithmetics and comparison below.